cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1302,7 +1304,8 @@ struct CopyFixedWidth<BooleanType> {
 };
 
 template <typename Type>
-struct CopyFixedWidth<Type, enable_if_number<Type>> {
+struct CopyFixedWidth<
+    Type, enable_if_t<is_number_type<Type>::value || is_interval_type<Type>::value>> {
```

Comment:
Out of curiosity, why `is_interval_type` here while IfElseFunctor only allows month-day-nano intervals?

Reply:
This helper is used by coalesce/choose/case_when so this is part of adding support for all of those functions.

if_else should support all the interval types. It uses GenerateTypeAgnosticPrimitive, so DayTime and Month intervals are handled as integers.